File: platforms/android/src/main/java/com/amap/agenui/AGenUI.java
package com.amap.agenui;

import android.content.Context;

import androidx.annotation.Keep;
import androidx.annotation.RestrictTo;

import com.amap.agenui.function.IFunction;
import com.amap.agenui.function.PlatformFunction;
import com.amap.agenui.render.component.ComponentRegistry;
import com.amap.agenui.render.component.IComponentFactory;
import com.amap.agenui.render.font.FontRegistry;
import com.amap.agenui.render.image.ImageLoader;
import com.amap.agenui.render.image.ImageLoaderConfig;
import com.amap.agenui.render.surface.ThemeException;
import com.amap.agenui.render.utils.AGenUILogger;

@Keep
public class AGenUI {
    private static final String TAG = "AGenUI";

    static {
        System.loadLibrary("amap_AGenUI");
    }

    private static volatile AGenUI sInstance = null;
    private static final Object sLock = new Object();

    private volatile long nativePtr = 0;
    private volatile boolean isInitialized = false;
    private volatile Context appContext = null;

    /**
     * Private constructor to prevent direct external instantiation
     */
    private AGenUI() {
    }

    /**
     * Returns the AGenUI singleton instance
     *
     * @return AGenUI singleton instance
     */
    public static AGenUI getInstance() {
        if (sInstance == null) {
            synchronized (sLock) {
                if (sInstance == null) {
                    sInstance = new AGenUI();
                }
            }
        }
        return sInstance;
    }

    /**
     * Initializes the AGenUI Engine
     *
     * Performs the following steps:
     * 1. Loads Native modules
     * 2. Creates the Engine instance (initAGenUIEngine)
     * 3. Initializes SkillManager and registers platform Skills
     *
     * @throws RuntimeException if initialization fails
     */
    public void initialize(Context applicationContext) {
        synchronized (sLock) {
            if (isInitialized) {
                AGenUILogger.w(TAG, "Module already initialized");
                return;
            }

            try {
                appContext = applicationContext.getApplicationContext();
                nativePtr = nativeInitAGenUIEngine();
                if (AGenUILogger.isLoggingEnabled()) {
                    AGenUILogger.i(TAG, "AGenUI Engine created: nativePtr=" + nativePtr);
                }
                ComponentRegistry.registerBuiltInComponents();

                isInitialized = true;
                AGenUILogger.i(TAG, "AGenUI Engine initialized successfully");
            } catch (Exception e) {
                AGenUILogger.e(TAG, "Failed to initialize AGenUI Engine", e);
                throw new RuntimeException("Failed to initialize AGenUI Engine", e);
            }
        }
    }
    
    /**
     * Sets a custom logger delegate to receive log callbacks from the engine.
     * 
     * This method should be called BEFORE initialize() to ensure all engine logs
     * are captured. If called after initialization, only subsequent logs will
     * use the custom delegate.
     * 
     * Example usage:
     * <pre>
     * AGenUI.getInstance().setCustomLogger(new IAGenUILogger() {
     *     {@literal @}Override
     *     public void onLog(int level, String tag, String func, int line, String message) {
     *         // Custom logging implementation
     *         AGenUILogger.d(tag, "[" + func + "@" + line + "] " + message);
     *     }
     * });
     * AGenUI.getInstance().initialize(context);
     * </pre>
     * 
     * @param customLogger Custom logger implementation. Pass null to use default logging.
     */
    public void setCustomLogger(IAGenUILogger customLogger) {
        if (!isInitialized()) {
            AGenUILogger.w(TAG, "setCustomLogger: Engine not initialized");
            return;
        }

        AGenUILogger.getInstance().setCustomLogger(customLogger);
    }

    /**
     * Set the minimum log level. Messages with level below this threshold are filtered out
     * on both the Java path and the C++ engine path (via IRuntimeLogger::getMinLevel()),
     * so filtered levels skip variadic formatting entirely.
     *
     * @param level One of IAGenUILogger LEVEL_DEBUG(0) ... LEVEL_PERFORMANCE(5). Out-of-range
     *              values fall back to LEVEL_DEBUG (no filtering).
     */
    public void setMinLogLevel(int level) {
        AGenUILogger.getInstance().setMinLogLevel(level);
    }

    /**
     * @return The currently configured minimum log level.
     */
    public int getMinLogLevel() {
        return AGenUILogger.getInstance().getMinLogLevel();
    }

    /**
     * Checks whether the Engine has been initialized
     *
     * @return true if the Engine is initialized
     */
    public boolean isInitialized() {
        return isInitialized && nativePtr != 0;
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public Context getApplicationContextForSdk() {
        return appContext;
    }

    /**
     * Creates a SurfaceManager instance
     *
     * @return instanceId (instance identifier)
     * @throws IllegalStateException if the Engine is not initialized or the native layer fails to create one
     */
    public int createSurfaceManager() throws IllegalStateException {
        if (!isInitialized()) {
            throw new IllegalStateException("createSurfaceManager: AGenUI engine is not initialized");
        }
        int instanceId = nativeCreateSurfaceManager();
        if (instanceId == 0) {
            throw new IllegalStateException("createSurfaceManager: native call failed");
        }

        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.i(TAG, "SurfaceManager created: instanceId=" + instanceId);
        }
        return instanceId;
    }

    /**
     * Destroys a SurfaceManager instance
     *
     * @param instanceId The instanceId of the SurfaceManager to destroy
     */
    public void destroySurfaceManager(int instanceId) {
        if (!isInitialized()) {
            AGenUILogger.w(TAG, "destroySurfaceManager: Engine not initialized");
            return;
        }
        nativeDestroySurfaceManager(instanceId);
        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.i(TAG, "SurfaceManager destroyed: engineId=" + instanceId);
        }
    }


    private boolean isConfigValid(String methodName, String config) {
        if (config == null || config.isEmpty()) {
            if (AGenUILogger.isLoggingEnabled()) {
                AGenUILogger.w(TAG, methodName + ": config is null or empty");
            }
            return false;
        }
        if (!isInitialized()) {
            AGenUILogger.e(TAG, methodName + ": Engine not initialized");
            return false;
        }
        return true;
    }

    private boolean loadThemeConfig(String themeConfig) {
        if (!isConfigValid("loadThemeConfig", themeConfig)) {
            return false;
        }
        return nativeLoadThemeConfig(themeConfig);
    }

    private boolean loadDesignTokenConfig(String designTokenConfig) {
        if (!isConfigValid("loadDesignTokenConfig", designTokenConfig)) {
            return false;
        }
        return nativeLoadDesignTokenConfig(designTokenConfig);
    }

    /**
     * Sets the day/night mode
     *
     * @param mode Mode value: "light" or "dark"
     */
    public void setDayNightMode(String mode) {
        if (mode == null || mode.isEmpty()) {
            AGenUILogger.w(TAG, "setDayNightMode: mode is null or empty");
            return;
        }
        if (!isInitialized()) {
            AGenUILogger.e(TAG, "setDayNightMode: Engine not initialized");
            return;
        }
        nativeSetDayNightMode(mode);
    }

    /**
     * Sets whether the host app is a debug build so the SDK can adjust its internal behavior.
     *
     * Should be called right after initialize().
     *
     * @param isDebug true for debug builds, false for release.
     *                Defaults to false if never called.
     */
    public void setDebug(boolean isDebug) {
        if (!isInitialized()) {
            AGenUILogger.e(TAG, "setDebug: Engine not initialized");
            return;
        }
        nativeSetDebug(isDebug);
    }

    /**
     * Gets whether the host app is a debug build, previously set via {@link #setDebug(boolean)}.
     *
     * @return true for debug builds. Returns false if never set or engine not initialized.
     */
    public boolean isDebug() {
        if (!isInitialized()) {
            AGenUILogger.w(TAG, "isDebug: Engine not initialized, returning default");
            return false;
        }
        return nativeIsDebug();
    }

    /**
     * Sets path configuration for the engine
     *
     * @param configJson Path configuration JSON string
     *                   Supported keys: "templateDir" - absolute path to the template directory
     * @return true if configuration was applied successfully, false otherwise
     */
    public boolean setPathConfig(String configJson) {
        if (!isConfigValid("setPathConfig", configJson)) {
            return false;
        }
        return nativeSetPathConfig(configJson);
    }

    /**
     * Registers the default theme configuration
     * <p>
     * Registers both the theme JSON and DesignToken JSON simultaneously.
     *
     * @param theme       Theme configuration JSON string
     * @param designToken DesignToken configuration JSON string
     * @throws ThemeException if registration fails
     */
    public void registerDefaultTheme(String theme, String designToken) throws ThemeException {
        boolean themeOk = loadThemeConfig(theme);
        if (!themeOk) {
            throw new ThemeException("Failed to register theme config");
        }
        boolean tokenOk = loadDesignTokenConfig(designToken);
        if (!tokenOk) {
            throw new ThemeException("Failed to register design token config");
        }
        AGenUILogger.i(TAG, "✓ Default theme registered successfully");
    }


    public void registerFunction(IFunction function) {
        nativeRegisterFunction(
                function.getConfig().getName(),
                function.getConfig().toJSON(),
                new PlatformFunction(function));
    }

    public void unregisterFunction(String name) {
        nativeUnregisterFunction(name);
    }

    /**
     * Declares a component property as a container of dynamic values.
     * <p>
     * By default a property value is parsed but not descended into, so a nested object
     * or array is stored verbatim and any {@code {"path":...}} binding or
     * {@code {"call":...}} function call inside it never resolves. Registering the
     * property makes the engine walk its entire subtree instead, resolving nested
     * dynamic values at any depth — including bindings inside a function call's
     * arguments, and relative paths inside a List item template.
     * <p>
     * Example — {@code AmapText} carries its rich-text runs in a {@code spans} array,
     * where each run has its own bound text and token-resolved colors:
     * <pre>{@code
     * AGenUI.getInstance().registerComponent("AmapText", new AmapA2UITextComponentFactory());
     * AGenUI.getInstance().registerDeepParseProperty("AmapText", "spans");
     * }</pre>
     *
     * @param componentType Component type, i.e. the JSON {@code component} field
     *                      (e.g. "AmapText") — applies to every instance of that type
     * @param propertyName  Property name (e.g. "spans")
     * @return true on success, false when either name is empty
     * @implNote Must be called before the first render. A declaration made afterwards
     *         does not affect components that have already been parsed.
     */
    public boolean registerDeepParseProperty(String componentType, String propertyName) {
        if (!isConfigValid("registerDeepParseProperty", componentType)
                || !isConfigValid("registerDeepParseProperty", propertyName)) {
            return false;
        }
        return nativeRegisterDeepParseProperty(componentType, propertyName);
    }



    /**
     * Registers a custom component factory
     * <p>
     * If the component type already exists, it will be overwritten. Takes effect immediately
     * after registration and is shared across all Surfaces.
     *
     * @param type    Component type (e.g. "MyCustomCard")
     * @param creator Component factory instance
     */
    public void registerComponent(String type, IComponentFactory creator) {
        if (type == null || type.isEmpty() || creator == null) {
            AGenUILogger.w(TAG, "registerComponent: invalid parameters");
            return;
        }
        ComponentRegistry.registerComponent(type, creator);
        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.i(TAG, "registerComponent: type=" + type);
        }
    }

    public void unregisterComponent(String type) {
        if (type == null || type.isEmpty()) {
            AGenUILogger.w(TAG, "unregisterComponent: invalid parameters");
            return;
        }
        ComponentRegistry.unregisterComponent(type);
        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.i(TAG, "unregisterComponent: type=" + type);
        }
    }


    /**
     * Registers a global image loader
     * <p>
     * All image components will use this loader to load network images.
     *
     * @param loader ImageLoader instance
     */
    public void registerImageLoader(ImageLoader loader) {
        if (loader == null) {
            AGenUILogger.w(TAG, "registerImageLoader: loader is null");
            return;
        }
        ImageLoaderConfig.getInstance().setLoader(loader);
        AGenUILogger.i(TAG, "registerImageLoader: success");
    }

    /**
     * Registers a custom font from an absolute file path.
     *
     * <p>The font becomes available immediately for {@code font-family} CSS
     * property resolution in Text, RichText and TextField components.
     *
     * @param familyName the family name to reference in {@code font-family}
     * @param filePath   absolute path to a {@code .ttf} or {@code .otf} file
     * @return true if registration succeeds
     */
    public boolean registerFont(String familyName, String filePath) {
        if (familyName == null || familyName.isEmpty() || filePath == null || filePath.isEmpty()) {
            AGenUILogger.w(TAG, "registerFont: familyName or filePath is null/empty");
            return false;
        }
        boolean ok = FontRegistry.getInstance().registerFont(familyName, filePath);
        if (ok) {
            AGenUILogger.i(TAG, "registerFont: registered '" + familyName + "' from file");
        } else {
            AGenUILogger.e(TAG, "registerFont: failed to load font from " + filePath);
        }
        return ok;
    }

    /**
     * Registers a custom font from the application's assets directory.
     *
     * <p>The font becomes available immediately for {@code font-family} CSS
     * property resolution in Text, RichText and TextField components.
     *
     * @param familyName the family name to reference in {@code font-family}
     * @param assetPath  path relative to the assets directory
     *                   (e.g. {@code "fonts/MyFont.ttf"})
     * @return true if registration succeeds
     */
    public boolean registerFontFromAsset(String familyName, String assetPath) {
        if (familyName == null || familyName.isEmpty() || assetPath == null || assetPath.isEmpty()) {
            AGenUILogger.w(TAG, "registerFontFromAsset: familyName or assetPath is null/empty");
            return false;
        }
        if (appContext == null) {
            AGenUILogger.e(TAG, "registerFontFromAsset: engine not initialized");
            return false;
        }
        boolean ok = FontRegistry.getInstance().registerFontFromAsset(familyName, assetPath, appContext.getAssets());
        if (ok) {
            AGenUILogger.i(TAG, "registerFontFromAsset: registered '" + familyName + "' from assets");
        } else {
            AGenUILogger.e(TAG, "registerFontFromAsset: failed to load font from " + assetPath);
        }
        return ok;
    }


    /**
     * Returns the AGenUI SDK version number
     *
     * @return SDK version number
     */
    public static String getVersion() {
        return nativeGetVersion();
    }


    /**
     * Returns the Native Engine pointer
     *
     * @return Native Engine pointer
     * @throws IllegalStateException if the Engine is not initialized
     */
    public long getNativePtr() {
        if (!isInitialized()) {
            throw new IllegalStateException("Engine not initialized");
        }
        return nativePtr;
    }

    /**
     * Destroys the Engine and releases all Native resources.
     * This method should be called when the application exits.
     */
    public void destroy() {
        synchronized (sLock) {
            if (!isInitialized) {
                AGenUILogger.w(TAG, "Engine not initialized, nothing to destroy");
                return;
            }

            try {
                if (nativePtr != 0) {
                    nativeDestroyAGenUIEngine();
                    AGenUILogger.i(TAG, "Engine destroyed successfully");
                }
            } catch (Exception e) {
                AGenUILogger.e(TAG, "Error destroying engine", e);
            } finally {
                nativePtr = 0;
                isInitialized = false;
                sInstance = null;
            }
        }
    }

    private native long nativeInitAGenUIEngine();
    private native void nativeDestroyAGenUIEngine();

    public static native String nativeGetVersion();
    public static native int nativeCreateSurfaceManager();
    public static native void nativeDestroySurfaceManager(int instanceId);

    public static native boolean nativeSetPathConfig(String configJson);

    private static native boolean nativeLoadThemeConfig(String themeConfig);
    private static native boolean nativeLoadDesignTokenConfig(String designTokenConfig);
    private static native void nativeSetDayNightMode(String mode);
    private static native void nativeSetDebug(boolean isDebug);
    private static native boolean nativeIsDebug();

    public static native void nativeRegisterFunction(String name, String config, Object function);
    public static native void nativeUnregisterFunction(String name);
    private static native boolean nativeRegisterDeepParseProperty(String componentType, String propertyName);
    public static native void nativeOnAsyncCallbackResult(long callbackPtr, int status, String data, String error);

    /**
     * Parses a CSS color value string (solid color or gradient).
     *
     * @param cssValue CSS color string, e.g. "red", "#ff0000", "linear-gradient(...)"
     * @return Parsed ColorValue object, or null if parsing fails
     */
    public static native ColorValue nativeParseColor(String cssValue);

    /**
     * Parses a CSS edge insets shorthand string (margin / padding / inset etc.).
     *
     * @param cssValue CSS shorthand string, e.g. "10px", "10px 20%", "10px 20px 30px 40px"
     * @return Parsed EdgeInsetsValue object, or null if parsing fails
     */
    public static native EdgeInsetsValue nativeParseEdgeInsets(String cssValue);
}

File: platforms/ios/AGenUI/Classes/AGenUI.swift
//
//  AGenUI.swift
//  AGenUI
//
// Created on 2026/3/18.
//

import Foundation

/// AGenUI SDK global entry point
///
/// Singleton responsible for global registration of theme, DesignToken, day/night mode, skills, components, and image loaders.
@objc(AGenUISDK) public class AGenUISDK: NSObject {

    private override init() {
        super.init()
    }

    // MARK: - Engine Bridge

    private static var engineBridge: AGenUIEngineBridge {
        return AGenUIEngineBridge.sharedInstance()
    }

    /// Strong references to registered functions to prevent deallocation
    private static var registeredFunctions: [String: Function] = [:]
    private static let fuctionsLock = NSLock()

    // MARK: - Global Configuration

    // MARK: - Theme Management

    /// Register default theme configuration
    ///
    /// Registers theme and design token configurations to customize UI component appearance.
    ///
    /// - Parameters:
    ///   - theme: Theme configuration JSON string
    ///   - designToken: Design token configuration JSON string
    /// - Returns: AGenUIError with result and message fields
    @objc public static func registerDefaultTheme(_ theme: String, designToken: String) -> AGenUIError {
        Logger.shared.debug("registerDefaultTheme - theme length: \(theme.count), designToken length: \(designToken.count)")
        
        // 1. Wrap theme with "default" key
        guard let themeData = theme.data(using: .utf8),
              let themeObject = try? JSONSerialization.jsonObject(with: themeData) else {
            Logger.shared.error("Failed to parse theme JSON")
            return AGenUIError(result: false, message: "Invalid theme JSON format")
        }
        
        let wrappedTheme: [String: Any] = ["default": themeObject]
        guard let wrappedData = try? JSONSerialization.data(withJSONObject: wrappedTheme),
              let wrappedThemeJson = String(data: wrappedData, encoding: .utf8) else {
            Logger.shared.error("Failed to serialize wrapped theme JSON")
            return AGenUIError(result: false, message: "Failed to wrap theme JSON")
        }
        
        // 2. Register theme
        let themeResult = engineBridge.loadThemeConfig(wrappedThemeJson)
        if !themeResult {
            Logger.shared.error("Theme registration failed")
            return AGenUIError(result: false, message: "Theme registration failed")
        }
        
        // 3. Register DesignToken
        let designTokenResult = engineBridge.loadDesignTokenConfig(designToken)
        if !designTokenResult {
            Logger.shared.error("DesignToken registration failed")
            return AGenUIError(result: false, message: "DesignToken registration failed")
        }
        
        Logger.shared.info("Default theme registered successfully")
        return AGenUIError(result: true, message: "Success")
    }

    /// Set day/night mode
    ///
    /// - Parameter mode: "light" or "dark"
    @objc public static func setDayNightMode(_ mode: String) {
        Logger.shared.debug("setDayNightMode: \(mode)")
        guard mode == "light" || mode == "dark" else {
            Logger.shared.warning("setDayNightMode: invalid mode '\(mode)', expected 'light' or 'dark'")
            return
        }
        engineBridge.setDayNightMode(mode)
        Logger.shared.info("Day/Night mode set to: \(mode)")
    }

    /// Set whether the host app is a debug build so the SDK can adjust its internal behavior
    ///
    /// Should be called right after initialize(). Defaults to false if never called.
    ///
    /// - Parameter isDebug: true for debug builds, false for release
    @objc public static func setDebug(_ isDebug: Bool) {
        Logger.shared.debug("setDebug: \(isDebug)")
        engineBridge.setDebug(isDebug)
        Logger.shared.info("Build type set to: \(isDebug ? "debug" : "release")")
    }

    /// Get whether the host app is a debug build, previously set via setDebug
    ///
    /// - Returns: true for debug builds. Returns false if never set.
    @objc public static func isDebug() -> Bool {
        return engineBridge.isDebug()
    }

    // MARK: - Path Configuration

    /// Set path configuration
    ///
    /// - Parameter configJson: Path configuration JSON string, e.g. {"templateDir": "/path/to/templates"}
    /// - Returns: AGenUIError with result and message fields
    @objc public static func setPathConfig(_ configJson: String) -> AGenUIError {
        Logger.shared.debug("setPathConfig - configJson length: \(configJson.count)")
        let result = engineBridge.setPathConfig(configJson)
        if result {
            Logger.shared.info("Path config set successfully")
            return AGenUIError(result: true, message: "Success")
        } else {
            Logger.shared.error("Failed to set path config")
            return AGenUIError(result: false, message: "Failed to parse path config JSON")
        }
    }

    // MARK: - Component Registration

    /// Register a custom component factory
    ///
    /// - Parameters:
    ///   - type: Component type name
    ///   - creator: Factory closure to create the component
    /// - Note: Overwrites if the component type already exists
    public static func registerComponent<T: Component>(_ type: String, creator: @escaping (String, [String: Any]) -> T) {
        Logger.shared.debug("registerComponent: \(type)")
        ComponentRegister.shared.register(type, creator: creator)
        Logger.shared.info("Custom component registered: \(type)")
    }
    
    /// Register a custom component factory (Objective-C compatible)
    ///
    /// - Parameters:
    ///   - type: Component type name
    ///   - creator: Factory closure to create the component, returns a Component instance
    /// - Note: Overwrites if the component type already exists
    @objc(registerComponent:creator:)
    public static func registerComponentObjC(_ type: String, creator: @escaping (String, [String: Any]) -> Component) {
        Logger.shared.debug("registerComponent (ObjC): \(type)")
        ComponentRegister.shared.register(type, creator: creator)
        Logger.shared.info("Custom component registered (ObjC): \(type)")
    }
    
    /// Register a custom component factory with explicit class name (Objective-C compatible)
    ///
    /// When called from ObjC, Swift generic T is erased to Component.
    /// This method accepts an explicit class name string to resolve the actual
    /// Component subclass for correct measurement dispatch.
    ///
    /// - Parameters:
    ///   - type: Component type name
    ///   - componentClassName: Fully qualified class name string for the Component subclass
    ///   - creator: Factory closure to create the component
    /// - Note: Overwrites if the component type already exists
    @objc(registerComponent:componentClassName:creator:)
    public static func registerComponentObjC(_ type: String, componentClassName: String, creator: @escaping (String, [String: Any]) -> Component) {
        Logger.shared.debug("registerComponent (ObjC): \(type), class: \(componentClassName)")
        ComponentRegister.shared.register(type, componentClassName: componentClassName, creator: creator)
        Logger.shared.info("Custom component registered (ObjC): \(type)")
    }
    
    @objc public static func unRegisterComponent(_ type: String) {
        Logger.shared.debug("unregisterComponent: \(type)")
        ComponentRegister.shared.unregister(type)
        Logger.shared.info("Custom component unregistered: \(type)")
    }

    /// Declare a component property as a container of dynamic values
    ///
    /// By default a property value is parsed but not descended into, so a nested object
    /// or array is stored verbatim and any `{"path":...}` binding or `{"call":...}`
    /// function call inside it never resolves. Registering the property makes the engine
    /// walk its whole subtree, resolving nested dynamic values at any depth — including
    /// bindings inside a function call's arguments, and relative paths inside a List
    /// item template.
    ///
    /// Example — AmapText carries its rich-text runs in a `spans` array:
    /// `AGenUI.registerDeepParseProperty("AmapText", propertyName: "spans")`
    ///
    /// Must be called before the first render; a declaration made later has no effect on
    /// components that have already been parsed.
    ///
    /// - Parameters:
    ///   - componentType: Component type, i.e. the JSON `component` field
    ///   - propertyName: Property name
    /// - Returns: true on success, false when either name is empty
    @discardableResult
    @objc(registerDeepParseProperty:propertyName:)
    public static func registerDeepParseProperty(_ componentType: String, propertyName: String) -> Bool {
        Logger.shared.debug("registerDeepParseProperty: \(componentType).\(propertyName)")
        let result = engineBridge.registerDeepParseProperty(forComponentType: componentType, propertyName: propertyName)
        if result {
            Logger.shared.info("Deep parse property registered: \(componentType).\(propertyName)")
        } else {
            Logger.shared.error("Failed to register deep parse property: \(componentType).\(propertyName)")
        }
        return result
    }

    // MARK: - Image Loader Registration

    /// Register a global image loader
    ///
    /// - Parameter loader: A loader instance implementing the ImageLoader protocol
    @objc(registerImageLoader:)
    public static func registerImageLoader(_ loader: ImageLoader) {
        Logger.shared.debug("registerImageLoader: \(type(of: loader))")
        ImageLoaderConfiguration.shared.loader = loader
        Logger.shared.info("ImageLoader registered: \(type(of: loader))")
    }

    // MARK: - Font Registration

    /// Register a custom font from a file URL.
    ///
    /// The font is loaded via `CTFontManager` and becomes available for
    /// `font-family` CSS property resolution in all text components.
    ///
    /// - Parameters:
    ///   - familyName: the name authors use in `font-family` (case-insensitive)
    ///   - fileURL: URL pointing to a `.ttf` or `.otf` font file
    /// - Returns: `true` if registration succeeds
    @objc(registerFont:fileURL:)
    public static func registerFont(_ familyName: String, fileURL: URL) -> Bool {
        let ok = FontRegistry.shared.registerFont(familyName: familyName, fileURL: fileURL)
        if ok {
            Logger.shared.info("registerFont: '\(familyName)' from \(fileURL.lastPathComponent)")
        } else {
            Logger.shared.error("registerFont: failed for '\(familyName)'")
        }
        return ok
    }

    /// Register a font name already available in the system.
    ///
    /// Use this when the font is declared in `Info.plist` under `UIAppFonts`
    /// and you want to expose it under a custom CSS family name.
    ///
    /// - Parameters:
    ///   - familyName: the name authors use in `font-family` (case-insensitive)
    ///   - fontName: the actual font name (PostScript name or family name)
    /// - Returns: `true` on success
    @objc(registerFont:fontName:)
    public static func registerFont(_ familyName: String, fontName: String) -> Bool {
        let ok = FontRegistry.shared.registerFont(familyName: familyName, fontName: fontName)
        if ok {
            Logger.shared.info("registerFont: '\(familyName)' -> '\(fontName)'")
        }
        return ok
    }

    // MARK: - FunctionCall Management

    /// Register a platform function with the engine
    ///
    /// - Parameter function: Object implementing the AGenUIFunctionProtocol
    @objc(registerFunction:)
    public static func registerFunction(_ function: Function) {
        let config = function.functionConfig
        let name = config.name
        guard !name.isEmpty else {
            Logger.shared.error("registerFunction: function name is empty")
            return
        }
        
        let configJson = config.toJSON()
        Logger.shared.debug("registerFunction - name: \(name)")
        
        // Create bridge callback that delegates to the function instance
        let bridgeCallback: AGenUIFunctionCallCallback = { [weak function] instanceId, surfaceId, argsJson in
            
            guard let function = function else {
                return "{\"status\":\"Error\",\"error\":\"Function deallocated\"}"
            }
            
            let context = FunctionCallContext(instanceId: Int(instanceId), surfaceId: surfaceId ?? "")
            let result = function.execute(context: context, params: argsJson)
            
            let resultDict: [String: Any] = [
                "status": result.result ? "Success" : "Error",
                "data": result.value
            ]
            
            guard let resultData = try? JSONSerialization.data(withJSONObject: resultDict),
                  let resultJson = String(data: resultData, encoding: .utf8) else {
                return "{\"status\":\"Error\",\"error\":\"Failed to serialize result\"}"
            }
            
            return resultJson
        }
        
        // Retain the function object to prevent deallocation
        fuctionsLock.lock()
        registeredFunctions[name] = function
        fuctionsLock.unlock()
        
        engineBridge.registerFunction(name, config: configJson, callback: bridgeCallback)
        Logger.shared.info("FunctionCall registered: \(name)")
    }

    /// Unregister a previously registered platform function
    ///
    /// - Parameter name: Function name to unregister
    @objc(unregisterFunctionWithName:)
    public static func unregisterFunction(_ name: String) {
        guard !name.isEmpty else {
            Logger.shared.error("unregisterFunction: name is empty")
            return
        }
        
        fuctionsLock.lock()
        registeredFunctions.removeValue(forKey: name)
        fuctionsLock.unlock()
        engineBridge.unregisterFunction(name)
        Logger.shared.info("FunctionCall unregistered: \(name)")
    }
    
    // MARK: - Logging
    
    /// Custom Logger
    ///
    /// - Parameter customLogger: Custom Logger
    /// - Note: By setting the delegate, other modules like can receive log output
    @objc public static func setCustomLogger(_ customLogger: LoggerDelegate?) {
        Logger.shared.delegate = customLogger
        // Wire the C++ engine to route runtime logs through the bridge so the
        // IRuntimeLogger extension (getMinLevel + delegate forwarding) actually takes effect.
        engineBridge.setRuntimeLogEnabled(true)
    }

    /// Set the minimum log level for AGenUI SDK.
    ///
    /// Only logs with level >= the specified level will be output.
    /// This controls both the Swift-side filtering and the C++ hot-path
    /// early-exit (via `IRuntimeLogger.getMinLevel()`), so filtered-out
    /// messages skip variadic formatting and cross-language bridging entirely.
    ///
    /// - Parameter level: The minimum log level to output.
    ///   Possible values: `.debug`(0), `.info`(1), `.warning`(2), `.error`(3), `.fatal`(4), `.performance`(5)
    @objc public static func setMinLogLevel(_ level: Logger.Level) {
        Logger.shared.minimumLevel = level
    }

    /// Get the current minimum log level for AGenUI SDK.
    ///
    /// - Returns: The current minimum log level.
    @objc public static func getMinLogLevel() -> Logger.Level {
        return Logger.shared.minimumLevel
    }

    // MARK: - Version Info

    /// Get the AGenUI SDK version
    ///
    /// - Returns: Version string
    @objc public static func getVersion() -> String {
        return AGenUIEngineBridge.sdkVersion()
    }
}

// MARK: - AGenUIError

@objc public class AGenUIError: NSObject {
    /// Whether the operation succeeded
    @objc public let result: Bool
    /// Error message
    @objc public let message: String
    
    @objc public init(result: Bool, message: String) {
        self.result = result
        self.message = message
    }
}


